### DISCUSSION TITLE
Couple of suggestions

### DISCUSSION THREAD

**Discussion by SonatasTaros** [TagStudioDev/TagStudio#348] [Ideas]
First off, I've been looking for a tagging system for Windows for a couple of decades now and even though this is an alpha release, it's already easily the best I've tried.

A couple of suggestions for some point in the future. I appreciate that most of these will probably already be in the "this is how it'll be intended to work when it's released" pile.

Firstly, it'd be nice if there was a page were *all* tags were listed. The "add tags" box seems to cap out at around 25-30, which means that you have to remember what tags you've actually got.

Secondly, it'd be good to have some heirarchy in that. At the moment everything is all in together. I think it'd make sense for the first tags listed to all be tags which don't have any subtabs (suggesting they're "top-level" tags), and if you click on them you get an indented list of tags that the top-level tag is a subtag in.  You could still have all the tags in the one view/list, just have the top-level tags always at the top of the list.

Thirdly, it'd be nice to have fuzzy search, at least as an option. At the moment if you've got a tag called, say, "In the Wild" and you search for "Wild" it won't show up. You have to search for the start of the tag's name.

Fourthly, remove text from text fields, or have it disappear when you start typing. At the moment you have to manually delete the "New Tab" text every time you make a new tab, and when you search for a tab it'll have your previous search still in the text box (although it won't show the results of that search unless you add or delete a letter). It'd be nice if that text disappeared when you started typing, rather than having to delete it first.

Fifthly, it'd be nice to have a keyboard shortcut to add tags. There's one for making tags, but you have to manually click the "add tag" button to add them to a file.

Sixthly, it'd be nice to be able to re-order subtags. It makes sense for the subtag at the top to be the one that appears in brackets in the tag's name, but that means that if you want something else there after you've already added subtags that you have to delete the ones you've added and then re-add them all.

Finally, it'd be nice to have the option for an "incognito mode". If you've got a shared computer there may be things you'd want to tag that you wouldn't want other people to be able to search for. Whether that's whole libraries that are protected in some way (password?) or tags or files, I don't know.

**Comment by samuellieberman** [comment]
Dang it, I typed out an entire response to this, but then lost it.

> Firstly, it'd be nice if there was a page were _all_ tags were listed. The "add tags" box seems to cap out at around 25-30, which means that you have to remember what tags you've actually got.

![tag_limit code](https://github.com/user-attachments/assets/c66cd889-8ffa-4e36-bfb5-19abae227efa)

This is because of the tag_limit attribute set for that window. The cap for the "Add Tags" window is 30. If you want to be able to see more tags, you should try the "Library Tags" window by clicking `Edit>Manage Tags`. "Library Tags" has a cap of 100.

I've suggested increasing these caps to CyanVoxel, but as I understand it, he is worried about the program lagging from having too many GUI elements loaded at one time. I believe that his goal is for the tags to dynamically load and unload as you scroll through the list.

> Secondly, it'd be good to have some hierarchy in that. At the moment everything is all in together. I think it'd make sense for the first tags listed to all be tags which don't have any subtabs (suggesting they're "top-level" tags), and if you click on them you get an indented list of tags that the top-level tag is a subtag in. You could still have all the tags in the one view/list, just have the top-level tags always at the top of the list.

You should add a feature request for that. Keep in mind that tags aren't strictly hierarchical. It is possible for one tag to inherit from multiple parents, or even to create inheritance loops.

Currently, if you wait for Alpha version 9.4, then #329 would allow you to choose which tags are grouped together in those panels by selecting their colors or editing their names. I have also worked one some more advanced tag sorting options in #336, but for now there aren't any immediate solutions to sorting tags in the main project.

I think you have made good suggestions, but I don't have much to comment on beyond these two.

**Comment by SonatasTaros** [comment]
> This is because of the tag_limit attribute set for that window. The cap for the "Add Tags" window is 30. If you want to be able to see more tags, you should try the "Library Tags" window by clicking Edit>Manage Tags. "Library Tags" has a cap of 100.

100's better, but I feel like the Add Tags window is where you will really want to be able to see them. And imagine you're doing something like wildlife photography, or building a sample library, or any number of things - it'll be really easy to get over 100 tags, very quickly. If there are ones that you don't use very frequently but which are still useful then it'd be very easy to forget to add that tag because you don't remember off the top of your head that it exists. And I think in many ways that a categorisation system that's incomplete is worse than not having one at all, because you'll tend to rely on it and will be more likely to miss things that you shouldn't miss because they're missing a tag.

> I've suggested increasing these caps to CyanVoxel, but as I understand it, he is worried about the program lagging from having too many GUI elements loaded at one time. I believe that his goal is for the tags to dynamically load and unload as you scroll through the list.

Oh, it should definitely be borne in mind that I know almost nothing about programming and I understand that there will be technical limitations and obstacles that I don't know about. These are just things that occurred to me while using it. I appreciate a) that they may not even be possible, and b) even if they are this isn't being developed *for me*.

> You should add a feature request for that.

I'm completely new to Github and, to be honest, I thought that's what I was doing with this thread.

> Keep in mind that tags aren't strictly hierarchical. It is possible for one tag to inherit from multiple parents, or even to create inheritance loops.

Oh, for sure. It's just that at the moment it seems like the use-case it's being written for is for people who know exactly what tags they've got and who will find them by searching. I think there could be utility in being able to navigate by clicking on tabs almost like folders (or Wikipedia pages, perhaps) and that showing you what other tabs are linked to them.

> Currently, if you wait for Alpha version 9.4, then https://github.com/TagStudioDev/TagStudio/pull/329 would allow you to choose which tags are grouped together in those panels by selecting their colors or editing their names. I have also worked one some more advanced tag sorting options in https://github.com/TagStudioDev/TagStudio/pull/336, but for now there aren't any immediate solutions to sorting tags in the main project.

Sorting by colour is something that I actually thought of, too. I know that a lot of the answers for things like this will be "wait, it's an alpha".

**Reply by samuellieberman** [reply]
> > I've suggested increasing these caps to CyanVoxel, but as I understand it, he is worried about the program lagging from having too many GUI elements loaded at one time. I believe that his goal is for the tags to dynamically load and unload as you scroll through the list.
> 
> Oh, it should definitely be borne in mind that I know almost nothing about programming and I understand that there will be technical limitations and obstacles that I don't know about. These are just things that occurred to me while using it. I appreciate a) that they may not even be possible, and b) even if they are this isn't being developed _for me_.

Personally I've already hit the 100 tag limit and I changed the code in my fork to increase the default limit to 1000. I would be surprised if I experienced any amount of lag before hitting 10,000 tags, but I don't think it's an issue that's much worth talking about at this time.

> > You should add a feature request for that.
> 
> I'm completely new to Github and, to be honest, I thought that's what I was doing with this thread.

That's understandable. Try clicking [here](https://github.com/TagStudioDev/TagStudio/issues/new/choose) to post in "Issues" instead of "Discussions." Good luck!

**Comment by Tishj** [comment]
I think all of these except for 4 and 7 are covered by the (upcoming) release of 9.5
